src/DataStructures/ModalVector.hpp

```diff
@@ -0,0 +1,485 @@
+    set_data_ref(rhs->data(), rhs->size_);
+  }
+  void set_data_ref(double* start, size_t size) noexcept {
+    size_ = size;
+    owned_data_ = decltype(owned_data_){};
+    data_ = decltype(data_){start, size_};
+    owning_ = false;
+  }
+  // @}
+
+  /// Returns true if the class owns the data
+  bool is_owning() const noexcept { return owning_; }
+
+  // @{
+  /// Access ith element
+  double& operator[](const size_type i) noexcept {
+    ASSERT(i < size_, "i = " << i << ", size = " << size_);
+    // clang-tidy: do not use pointer arithmetic
+    return data_[i];  // NOLINT
+  }
+  const double& operator[](const size_type i) const noexcept {
+    ASSERT(i < size_, "i = " << i << ", size = " << size_);
+    // clang-tidy: do not use pointer arithmetic
+    return data_[i];  // NOLINT
+  }
+  // @}
+
+  // @{
+  /// Access to the pointer
+  double* data() noexcept { return data_.data(); }
+  const double* data() const noexcept { return data_.data(); }
+  // @}
+
+  // @{
+  /// Returns iterator to beginning of data
+  decltype(auto) begin() noexcept { return data_.begin(); }
+  decltype(auto) begin() const noexcept { return data_.begin(); }
+  // @}
+  // @{
+  /// Returns iterator to end of data
+  decltype(auto) end() noexcept { return data_.end(); }
+  decltype(auto) end() const noexcept { return data_.end(); }
+  // @}
+
+  /// Serialization for Charm++
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  // @{
+  /// See the Blaze library documentation for details on these functions since
+  /// they merely forward to Blaze.
+  /// https://bitbucket.org/blaze-lib/blaze/overview
+  ModalVector& operator=(const double& rhs) noexcept {
+    data_ = rhs;
+    return *this;
+  }
+
+  ModalVector& operator+=(const ModalVector& rhs) noexcept {
+    data_ += rhs.data_;
+    return *this;
+  }
+  template <typename VT, bool VF>
+  ModalVector& operator+=(const blaze::Vector<VT, VF>& rhs) noexcept {
+    data_ += rhs;
+    return *this;
+  }
+  ModalVector& operator+=(const double& rhs) noexcept {
+    data_ += rhs;
+    return *this;
+  }
+
+  ModalVector& operator-=(const ModalVector& rhs) noexcept {
+    data_ -= rhs.data_;
+    return *this;
+  }
+  template <typename VT, bool VF>
+  ModalVector& operator-=(const blaze::Vector<VT, VF>& rhs) noexcept {
+    data_ -= rhs;
+    return *this;
+  }
+  ModalVector& operator-=(const double& rhs) noexcept {
+    data_ -= rhs;
+    return *this;
+  }
+
+  template <typename VT, bool VF>
+  ModalVector& operator*=(const blaze::Vector<VT, VF>& rhs) noexcept {
+    data_ *= rhs;
+    return *this;
+  }
+  ModalVector& operator*=(const double& rhs) noexcept {
+    data_ *= rhs;
+    return *this;
+  }
+  ModalVector& operator*=(const ModalVector& rhs) noexcept {
+    data_ *= rhs.data_;
+    return *this;
+  }
+  // FIXME PK: Can we avoid the instantiation of new DV object in this function
+  ModalVector& operator*=(const DataVector& rhs) noexcept {
```

Comment:
What is the use case for ModalVector-DataVector math?  I thought this was the sort of thing this class was created to avoid.

Reply:
The use case is for operations like filtering, where one may want to able to multiply a ModalVector by some other vector (a DataVector?) that represents the filter function.

Reply:
I guess I'm still confused about what all the different kinds of vector are for.  We now have two PRs adding new ones, and I don't understand how I decide, for a given quantity, which one I should use.

Reply:
ModalVector is *exclusively* to store spectral coefficients, with no non-arithmetic math overloaded for them. They are only allowed to interact with DataVector (limited arithmetic) to enable filtering, with the assumption that spectral filters will be stored in DataVectors.  We could require filters to be 'explicitly' constructed as ModalVectors as well, in which case ModalVector-DataVector math can be removed. Alternately, we can require filters to be DataVectors with dedicated tags (after an effective [tagging feature in blaze](https://bitbucket.org/blaze-lib/blaze/issues/166/allow-tagging-blaze-expressions) comes about). In this case, we can enable ModalVector-DataVector math only for this tag. The latter solution makes most sense to me.

Reply:
According to the wording of #696, filters should be stored as DenseVectors rather than as DataVectors, because DataVectors are for data over the computational grid (i.e. nodal data).

Reply:
Yes we should make the use cases for these data structures perfectly clear before adding them. My understanding is this:

- **DenseVector**: Generic collection of numbers, no particular meaning, used e.g. in the elliptic linear solver where it's just a matrix equation to solve. Just a thin wrapper around Blaze, so that we don't have to use anything from the `blaze` namespace directly.
- **DataVector**/**ModalVector**: Nodal/Modal coefficients **only**, so has a strong semantic meaning. Only implements functionality that makes sense for nodal/modal coefficients. Based on custom PointerVector implementation for efficiency.

I'll see to adding documentation to #696.